googlefirestore/src/main/java/com/yahoo/ycsb/db/GoogleFirestoreClient.java

```diff
@@ -0,0 +1,247 @@
+import com.google.cloud.firestore.SetOptions;
+import com.google.cloud.firestore.WriteResult;
+import com.yahoo.ycsb.ByteIterator;
+import com.yahoo.ycsb.DB;
+import com.yahoo.ycsb.DBException;
+import com.yahoo.ycsb.Status;
+import com.yahoo.ycsb.StringByteIterator;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.Vector;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * YCSB Client for Google's Cloud Firestore.
+ */
+public class GoogleFirestoreClient extends DB {
+  /**
+   * The names of properties which can be specified in the config files and flags.
+   */
+  public static final class GoogleFirestoreProperties {
+    private GoogleFirestoreProperties() {}
+
+    /**
+     * The Cloud Firestore project ID to use when running the YCSB benchmark.
+     */
+    static final String PROJECT = "googlefirestore.projectId";
+    static final String PRIVKEYFILE = "googlefirestore.serviceAccountKey";
+    static final String DEBUG = "googlefirestore.debug";
+  }
+
+  private static final Logger LOGGER = Logger.getLogger(GoogleFirestoreClient.class);
+
+  private static Firestore db;
+
+  @Override
+  public void init() throws DBException {
+    if (db != null) {
+      return;
+    }
+    Properties properties = getProperties();
+    String debug = properties.getProperty(GoogleFirestoreProperties.DEBUG);
+    if (null != debug && "true".equalsIgnoreCase(debug)) {
+      LOGGER.setLevel(Level.DEBUG);
+    }
+    String projectId = properties.getProperty(GoogleFirestoreProperties.PROJECT);
+    if (projectId == null) {
+      throw new DBException("Must provide project ID.");
+    }
+    String privateKeyFile = properties.getProperty(GoogleFirestoreProperties.PRIVKEYFILE);
+    if (privateKeyFile == null) {
+      throw new DBException("Must provide full path to private key file.");
+    }
+    try {
+      GoogleCredentials gCreds = GoogleCredentials.fromStream(new FileInputStream(privateKeyFile));
+      FirestoreOptions fsOptions = FirestoreOptions.newBuilder().setCredentials(gCreds).build();
+      db = fsOptions.getService();
+    } catch (FileNotFoundException e) {
+      LOGGER.error("Can't find key.", e);
+    } catch (IOException e) {
+      LOGGER.error("No file to import.", e);
+    }
+
+    LOGGER.info("Created Firestore client for project: "+ projectId);
+  }
+
+  @Override
+  public Status read(String table, String key, Set<String> fields, Map<String, ByteIterator> result) {
+    DocumentReference docRef = toReference(table, key);
+
+    try {
+      DocumentSnapshot docSs = docRef.get().get();
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug("read: Collection: " + table + " document: " + key);
+      }
+      // If document doesn't exist, will cause NPE.
```

Comment:
Is this comment correct? If we do a read on a key that doesn't exist we should return `Status.ERROR` rather than throw a NPE.

Reply:
removed the comment, as the code below it resolves issue.